add category selector to icon picker dialog

TayouVR opened the pull request:
it uses some regex shenanigans for this, probably not ideal, idk if theres a good way to filter the icons without adding extra metadata or storing them in subfolders

I think storing in subfolders would be ideal, but wouldn't work with old icons.


This was mostly a experiment from me to see if we can essentially have the KDE IconPicker dialog, but the only major difference is some minor changes in the delegate, and the categories. So I just went and added that

TayouVR committed:
add category selector to icon picker dialog

it uses some regex shenanigans for this, probably not ideal, idk if theres a good way to filter the icons without adding extra metadata or storing them in subfolders

Signed-off-by: Tayou <git@tayou.org>

TayouVR committed:
add category selector to icon picker dialog

it uses some regex shenanigans for this, probably not ideal, idk if theres a good way to filter the icons without adding extra metadata or storing them in subfolders

Signed-off-by: Tayou <git@tayou.org>

TayouVR committed:
rename IconPickerCategory and make public

Signed-off-by: Tayou <git@tayou.org>

TayouVR committed:
rename IconPickerCategory and make public

Signed-off-by: Tayou <git@tayou.org>

DioEgizio commented:
if you download a very old ftb legacy modpack (like magic world) those ones have their own ftb_logo_ftb icon that is downloaded and different from the in-launcher ones. That isn't shown in the "Modpacks" category

TayouVR commented:
yeah, I'm thinking maybe we can reorganize the builtin ones to make it clear what they are. I'd need to look into that.

TheKodeToad reviewed (commented):
I would suggest instead of using regex subclassing QSortProxyModel to override filterAcceptsRow and making use of the type field on MMCIcon

I would suggest this logic to determine the category:
- if type is builtin
  - if name() ends with "_legacy" category is "Builtin Legacy" otherwise "Builtin Modern"
- if type is not builtin
  - if name() starts with "curseforge_", "modrinth_" etc. the category is "Modpack" otherwise "Custom"

TayouVR commented:
that is way better, I wasn't sure how to do that initially, but got it working now. 

Looks like Legacy FTB packs don't have a prefix like `legacy_ftb_`.
I could filter for startsWith("FTB"), but idk
<img width="1226" height="261" alt="image" src="https://github.com/user-attachments/assets/eb426be9-097a-4fa6-b05d-f70c56146ab7" />

TayouVR committed:
use custom QSortFilterProxyModel impl

Signed-off-by: Tayou <git@tayou.org>

TayouVR committed:
use custom QSortFilterProxyModel impl

Signed-off-by: Tayou <git@tayou.org>

TheKodeToad commented:
Nice! Maybe a better solution for modpacks would be to start downloading them to a packs subfolder (we already support subfolders)

Edit: Oh yeah
> I think storing in subfolders would be ideal, but wouldn't work with old icons.

/shrug it seems like a better long term solution
Maybe we could just show every subfolder as a filter option?

TayouVR commented:
maybe. This isn't urgent, so I can look into that soon, sure.

I think subfolders would be good for this. ^^

Trial97 commented:
@TayouVR what is the status here? is this ready to review?

TayouVR commented:
I was actually looking into doing the folder thing, but our codebase is so convoluted, so ig maybe that can be done another time.

if the prefix based categories are acceptable to other maintainers, then it can go in as is.

TayouVR commented:
this has the approvals, and I think I wanna do some cleanup to instance creation before messing with the paths. I'll merge this.

TayouVR committed:
fix category filtering

I have no idea how this worked at all when I made the commit originally, but it works now, just as well as it did on the prior commit.

Further improvements, using subfolders and other metadata will be in another PR.
Signed-off-by: Tayou <git@tayou.org>
